### PR TITLE
Fix ACC_MODULE class verification

### PR DESCRIPTION
**Pull request by theresa-m**
* Class file is only a module if ACC_MODULE is the only flag set.
* Remove ACC_MODULE from CFR_CLASS_ACCESS_MASK. A valid class cannot have this flag set.

Fixes: #740 

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

Peer reviewer: @babsingh

### CODE REVIEW

File: runtime/bcutil/cfreader.c [hunk 1]
```diff
@@ -1174,7 +1174,7 @@ readPool(J9CfrClassFile* classfile, U_8* data, U_8* dataEnd, U_8* segment, U_8*
 				goto _errorFound;
 			}
 
-			if (!J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE)) {
+			if (classfile->accessFlags != CFR_ACC_MODULE) {
```
Comment (babsingh) [inline runtime/bcutil/cfreader.c]: generally, we do bit operations/checks on`accessFlags`. better to add this comment over here as well:
```
	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
	 * other flag in the access_flags item may be set. If other flags are set, then this
	 * file is not a module.
```

File: runtime/bcutil/cfreader.c [hunk 2]
```diff
@@ -2201,8 +2201,10 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 		goto _errorFound;
 	}
 
-	/* check annotations and interfaces for invalid flags */
-	if ((value & CFR_ACC_INTERFACE) && (value & ~CFR_INTERFACE_CLASS_ACCESS_MASK))
+	/* check annotations and interfaces for invalid flags. If the ACC_INTERFACE flag is set,
+	 * the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.
+	 */
+	if ((value & CFR_ACC_INTERFACE) && (value & CFR_INTERFACE_FORBIDDEN_ACCESS_MASK))
```
Comment (babsingh) [inline runtime/bcutil/cfreader.c]: better to use the macro: `J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE | CFR_INTERFACE_FORBIDDEN_ACCESS_MASK)`

File: runtime/bcutil/cfreader.c [hunk 3]
```diff
@@ -2442,8 +2445,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 		classfile->accessFlags |= CFR_ACC_ABSTRACT;
 	}
 
+	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
+	 * other flag in the access_flags item may be set. If other flags are set, then this
```
Comment (babsingh) [inline runtime/bcutil/cfreader.c]: `then no other flag in the access_flags item may be set`: may be -> should be?

**Comment by theresa-m** [comment]
Thanks @babsingh. I applied your changes.

File: runtime/bcutil/cfreader.c [hunk 4]
```diff
@@ -2442,8 +2449,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
...
 	}
 
+	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
+	 * other flag in the access_flags item should be set. If other flags are set, then this
+	 * file is not a module.
+	 */
 	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+			&& (classfile->majorVersion >= 53)
+			&& (classfile->accessFlags == CFR_ACC_MODULE)) {
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: I don't understand this change. It seems it should be checking if the ACC_MODULE bit is set as in the original code, but not if only the ACC_MODULE bit is set.

File: runtime/bcutil/cfreader.c [hunk 5]
```diff
@@ -2201,8 +2205,10 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 		goto _errorFound;
 	}
 
-	/* check annotations and interfaces for invalid flags */
-	if ((value & CFR_ACC_INTERFACE) && (value & ~CFR_INTERFACE_CLASS_ACCESS_MASK))
+	/* check annotations and interfaces for invalid flags. If the ACC_INTERFACE flag is set,
+	 * the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.
+	 */
+	if (J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE) && J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK))
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: ACC_MODULE is also invalid for Java 9

File: runtime/bcutil/cfreader.c [hunk 6]
```diff
@@ -2442,8 +2449,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 		classfile->accessFlags |= CFR_ACC_ABSTRACT;
 	}
 
+	/* Java 9 Spec: If the ACC_MODULE flag is set in the access_flags item, then no
+	 * other flag in the access_flags item should be set. If other flags are set, then this
+	 * file is not a module.
+	 */
 	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+			&& (classfile->majorVersion >= 53)
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: Is this check necessary after checking vmVersionShifted?

File: runtime/bcutil/cfreader.c (hunk 5 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Thanks. Looks like the Java 9 spec I've been using is a bit out of date.

File: runtime/bcutil/cfreader.c (hunk 4 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: In the Java 9 spec a file is considered a module only if ACC_MODULE is the only access flag set. I added this to double check that modules are not loaded like classes and to match Oracles behaviour for this scenario.

Comment (pshipton) [reply runtime/bcutil/cfreader.c]: Understood, but it seems you need to check if the module bit is set in the class, not if its the only bit set. Or is there there a check somewhere else that checks if the module bit is set? The code doesn't match the description in the comment.

File: runtime/bcutil/cfreader.c (hunk 6 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: These are not both needed.

File: runtime/bcutil/cfreader.c (hunk 4 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Ah the case where ACC_MODULE is set but its not an invalid module?

I also found the rule `If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set.` in the new spec. I will add a check for this in that should address that case.

**Comment by theresa-m** [comment]
Updated checks to java 9 spec:
* If ACC_INTERFACE is set, ACC_MODULE must not be set
* If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set

File: runtime/bcutil/cfreader.c [hunk 7]
```diff
@@ -2195,18 +2195,35 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 		goto _errorFound;
 	}
 
+	/* Java 9 spec: If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set. */
+	if ((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted) {
+		if ((value & CFR_ACC_MODULE) && (! (value & CFR_ACC_INTERFACE))) {
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: Since this is new code, please use the J9 BITS macros.

File: runtime/bcutil/cfreader.c [hunk 8]
```diff
@@ -2195,18 +2195,35 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
...
+	 * For Java 9 ACC_MODULE also must not be set
+	 */
+	 if (J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE))
 	{
-		errorCode = J9NLS_CFR_ERR_INTERFACE_CLASS__ID;
-		offset = endOfConstantPool;
-		goto _errorFound;
+		 if ((((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted)
+				 && J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK_9))
+			|| J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK)) {
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: Please move the `) {` to the next line for this multi-line if statement. It should line up with the if.

File: runtime/bcutil/cfreader.c (hunk 7 above)
Comment (pshipton) [reply runtime/bcutil/cfreader.c]: Doesn't this conflict with the below check that if CFR_ACC_INTERFACE is set then CFR_ACC_MODULE cannot be set?

File: runtime/bcutil/cfreader.c [hunk 9]
```diff
@@ -2442,8 +2460,7 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 		classfile->accessFlags |= CFR_ACC_ABSTRACT;
 	}
 
-	if ((vmVersionShifted >= BCT_Java9MajorVersionShifted)
-			&& (J9_ARE_ALL_BITS_SET(classfile->accessFlags, CFR_ACC_MODULE))) {
+	if ((classfile->majorVersion >= 53) && (classfile->accessFlags == CFR_ACC_MODULE)) {
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: Why the change to use classfile->majorVersion instead of vmVersionShifted? If its required, please add comments to explain it, and create or use an existing defined constant for 53. 

I still don't think its correct to check `accessFlags == CFR_ACC_MODULE` rather than checking if the CFR_ACC_MODULE bit is set.

**Comment by pshipton** [comment]
There used to be a check that .class'es with the module attribute must only have the ACC_MODULE bit set, which I thought was valid, but has since been removed.

**Comment by theresa-m** [comment]
I think I've just been making this worse by adding more checks - I checked that Hotspot is throwing the same error message for modules for all cases of a Java 9 class file with ACC_MODULE set, so in my latest push I simply moved the check further up in verification so it checks sooner, and removed ACC_MODULE from CFR_CLASS_ACCESS_MASK that I added in a fix earlier. Since ACC_MODULE is not part of a class that is valid I realized this doesn't make a lot of sense.

File: runtime/bcutil/cfreader.c (hunk 9 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Tests were failing in some cases using vmVersionShifted because a class file with version below 53 can be loaded by a Java 9 vm and this check should not apply.

File: runtime/bcutil/cfreader.c (hunk 7 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Yes - I'll these checks are correct. I've now simplified them into one check.

File: runtime/bcutil/cfreader.c [hunk 10]
```diff
@@ -2378,6 +2375,7 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	/* protect the callers with default values */
 	if ((flags & BCT_MajorClassFileVersionMask) == 0) {
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: I don't think this check is correct. It appears the MajorClassFileVersion portion of the flags is initialized further down in this function. It can be moved higher, but the classfile->majorVersion needs to be read in first. We shouldn't be initializing a default of Java 9.

Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Just pushed a change to address this.

File: runtime/bcutil/cfreader.c [hunk 11]
```diff
@@ -2130,11 +2130,16 @@ checkAttributes(J9CfrClassFile* classfile, J9CfrAttribute** attributes, U_32 att
...
+checkClassVersion(J9CfrClassFile* classfile, U_8* segment, U_32 vmVersionShifted)
 {
 	U_32 errorCode = J9NLS_CFR_ERR_MAJOR_VERSION__ID;
 	U_32 offset = 6;
-	U_16 max_allowed_version = flags >> BCT_MajorClassFileVersionMaskShift;
+	U_16 max_allowed_version = vmVersionShifted >> BCT_MajorClassFileVersionMaskShift;
+
+	/* protect the callers with default values */
+	if (0 == max_allowed_version) {
+		max_allowed_version = BCT_Java9MajorVersionShifted >> BCT_MajorClassFileVersionMaskShift;
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: This needs to be fixed not to default to anything. We need to consistently know which version of Java the VM is currently running, since it supports multiple versions.

File: runtime/bcutil/cfreader.c [hunk 12]
```diff
@@ -2171,11 +2176,7 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 		return -1;
 	}
 
-	if (vmVersionShifted >= BCT_Java9MajorVersionShifted) {
-		value = classfile->accessFlags & CFR_CLASS_ACCESS_MASK_9;
-	} else {
-		value = classfile->accessFlags & CFR_CLASS_ACCESS_MASK;
-	}
+	value = classfile->accessFlags & CFR_CLASS_ACCESS_MASK;
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: The vmVersionShifted parameter to the function is no longer needed.

Comment (pshipton) [reply runtime/bcutil/cfreader.c]: The vmVersionShifted parameter to the checkPool() function is no longer needed.

File: runtime/bcutil/cfreader.c (hunk 11 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Removed this check and fixed tests / cfdumper.

File: runtime/bcutil/cfreader.c [hunk 13]
```diff
@@ -2397,6 +2388,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 
 	NEXT_U16(classfile->minorVersion, index);
 	NEXT_U16(classfile->majorVersion, index);
+
+	/* Make sure that following verification uses the class file version number. VM version is
```
Comment (pshipton) [inline runtime/bcutil/cfreader.c]: Since this code was moved, it should change to say "the structure and static verification below" instead of "following verification"

File: runtime/oti/j9nonbuilder.h [hunk 14]
```diff
@@ -2017,6 +2017,7 @@ typedef struct J9BCTranslationData {
 #define BCT_InitInstanceFields  19
 #define BCT_ErrConstantPoolMapTooSmall  14
 #define BCT_InitCopyDoubles  5
+#define BCT_JavaMaxMajorVersionShifted BCT_Java10MajorVersionShifted
```
Comment (pshipton) [inline runtime/oti/j9nonbuilder.h]: Please flip the order so its 8, 9, 10, Max. Leave a blank line after Max to make it more noticable, and add a comment before 8 to say something like when adding a new major version update BCT_JavaMaxMajorVersionShifted to the maximum allowed value.

File: runtime/bcutil/cfreader.c (hunk 13 above)
Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Just updated the comment.
